docs/architecture/devops-for-aspnet-developers/actions-vs-pipelines.md

```diff
@@ -0,0 +1,118 @@
+
+Feature|GitHub|Azure Pipelines|Links
+--|--|--|--
+Hosted agents for public repos/projects|Free|[No free minutes](https://devblogs.microsoft.com/devops/change-in-azure-pipelines-grant-for-public-projects/) for public projects|[Azure Pipelines](/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml#capabilities-and-limitations) [GitHub](https://github.com/features/actions)
+Hosted agents for private repos/projects|2,000 minutes free per month, 3,000 minutes for Pro and Team licenses, 50,000 minutes for Enterprise license. Additional minutes may be purchased.|One free parallel job that can run for up to 60 minutes each time, until you've used 1,800 minutes (30 hours) per month. You can pay for additional capacity per parallel job. Paid parallel jobs remove the monthly time limit and allow you to run each job for up to 360 minutes (6 hours).
+Cross-platform|Yes|Yes|
+Scale set agents|No|Yes|Read about scale-set agents [here](/azure/devops/pipelines/agents/scale-set-agents?view=azure-devops)
+
+## Comparison of GitHub Actions and Azure Pipelines
+
+Azure Pipelines (YAML pipelines) provide a mature set of features such as approvals, artifact storage, deployment jobs, environments, gates, stages, templates, triggers, variable groups and more. For a full list of Azure Pipelines features, refer to the table [here](/azure/devops/pipelines/get-started/pipelines-get-started?view=azure-devops#feature-availability).
+
+GitHub Actions are evolving rapidly and provide a set of features such as triggers for almost all GitHub events, artifact storage, environments and environment rules, starter templates, matrices and more. Read more about the full feature set of GitHub Actions [here](https://docs.github.com/actions).
+
+### Feature comparison
+
+(As of March 2021)
+
+Feature|Description|GitHub Actions|Azure Pipelines
+--|--|--|--
+Approvals|Define approval conditions before moving further in the pipeline|Yes|Yes
+Artifacts|Upload, store and download artifacts from jobs|Yes|Yes
+Caching|Cache folders or files for subsequent runs|Yes|Yes
+Conditions|Specify conditions for steps or jobs|Yes|Yes
+Container Jobs|Run jobs inside a container|No|Yes
+Demands|Specify demands that must be met to match jobs to agents|No|Yes
+Dependencies|Specify dependencies between jobs or stages|Yes|Yes
+Deployment Groups|A logical set of target machines for deployments|No|Yes
+Deployment Jobs|Job that targets a deployment group|No|Yes
+Environments|A collection of resources to target or a logical environment|Yes|Yes
+Gates|Automatic collection and evaluation of signals to control continuation|No|Yes
+Jobs|Sequence of steps that are executed on an agent|Yes|Yes
+Service Containers|Manage the lifecycle of a containerized service instance availalbe for the duration of a job|Yes|Yes
+Service Connections|Abstract credentials to external systems|No|Yes
+Stages|Organize jobs in a pipeline|No|Yes
+Templates|Define reusable, parameterized building blocks|No|Yes
```

Comment:
```suggestion
Templates|Define reusable, parameterized building blocks|Yes|Yes
```

While they're a bit different than the traditional `Templates` offered by Azure Pipelines, based on the "Define reusable, parameterize building blocks" definition, it's easy to argue that GitHub Actions does have this feature capability. Isn't this the standard workflow composition UI, where GitHub knows the defined input and outputs?

Reply:
No, what I am going for is a group of steps that form a "complex" block that has parameters. You can do this in Azure Pipelines, but not in GHA. For example, if I want a template for all .NET core apps, I may want a workflow that checks out code, runs `dotnet build`, `dotnet test`, `dotnet publish` and publishes the output. If I want to repeat that over 5 repos, I have to copy/paste that workflow - I can't define the steps as a template with parameters. I can do that using Azure Pipelines.